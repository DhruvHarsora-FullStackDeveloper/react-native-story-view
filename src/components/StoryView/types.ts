export interface StoryViewProps {
  label?: string;
}
