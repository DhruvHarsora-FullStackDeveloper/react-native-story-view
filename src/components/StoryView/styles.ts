import { StyleSheet } from 'react-native';
import { Metrics, moderateScale, scale, verticalScale } from '../../theme';

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  scrollContainer: {
    alignItems: 'center',
    justifyContent: 'center',
  },
  divStory: {
    flexDirection: 'row',
    alignSelf: 'center',
  },
  imgStyle: {
    width: Metrics.screenWidth,
    height: Metrics.screenHeight,
    alignSelf: 'center',
    resizeMode: 'contain',
  },
  imageOverlay: {
    position: 'absolute',
    left: 0,
    right: 0,
    bottom: 0,
    top: 0,
    resizeMode: 'contain',
  },
  progressiveImageContainer: {
    backgroundColor: 'transparent',
  },
  parentView: {
    flexDirection: 'row',
  },
  customView: {
    position: 'absolute',
    flexDirection: 'column',
    width: Metrics.screenWidth,
  },
  topView: {
    position: 'absolute',
    flexDirection: 'column',
    width: Metrics.screenWidth,
    zIndex: 99,
  },
  bottomView: {
    position: 'absolute',
    bottom: 0,
    left: 0,
    right: 0,
    width: '100%',
    flexDirection: 'column',
    justifyContent: 'center',
    alignSelf: 'baseline',
    textAlignVertical: 'bottom',
    paddingTop: '3%',
    paddingBottom: '2%',
  },
  progressView: {
    flex: 1,
    width: '100%',
    position: 'absolute',
    flexDirection: 'row',
  },
  contentVideoView: {
    height: Metrics.screenHeight,
    width: Metrics.screenWidth,
    backgroundColor: 'transparent',
  },
  progressBarArray: {
    flexDirection: 'row',
    position: 'absolute',
    top: verticalScale(10),
    width: '98%',
    height: verticalScale(10),
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  progressBarContainer: {
    flex: 1,
    margin: moderateScale(2),
    borderRadius: verticalScale(10),
  },
  currentBarContainer: {
    position: 'absolute',
    top: 0,
    margin: 0,
  },
  //UserHeaderView
  userContainer: {
    flex: 1,
    justifyContent: 'center',
  },
  barUsername: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  image: {
    width: moderateScale(50),
    height: moderateScale(50),
    borderRadius: moderateScale(50),
    marginLeft: scale(8),
  },
  verifyIcon: {
    width: moderateScale(20),
    height: moderateScale(20),
    marginLeft: scale(5),
  },
  closeIcon: {
    width: moderateScale(25),
    height: moderateScale(25),
    marginRight: scale(8),
  },
  userView: {
    flexDirection: 'row',
    position: 'absolute',
    top: verticalScale(30),
    width: '98%',
    alignItems: 'center',
  },
  name: {
    fontSize: moderateScale(18),
    fontWeight: '500',
    marginLeft: 12,
    color: 'white',
  },
  time: {
    fontSize: moderateScale(12),
    fontWeight: '400',
    marginTop: verticalScale(3),
    marginLeft: scale(12),
    color: 'white',
  },
  loader: {
    alignItems: 'center',
    justifyContent: 'center',
  },
  loaderView: {
    flex: 1,
    position: 'absolute',
    top: '50%',
    left: '45%',
  },
});

export default styles;
