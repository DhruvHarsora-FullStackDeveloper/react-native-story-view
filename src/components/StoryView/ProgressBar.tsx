import React, { useEffect } from 'react';
import { Animated, LayoutRectangle, View } from 'react-native';
import { useProgressBar } from './hooks';
import styles from './styles';
import type { ProgressBarProps } from './types';

const ProgressBar = (props: ProgressBarProps) => {
  const { index, currentIndex } = props;

  const {
    barActiveColor,
    barInActiveColor,
    barHeight,
    scale,
    setWidth,
    setPauseTime,
    startTime,
    setStartTime,
  } = useProgressBar(props);

  const onLayoutAdded = (evt: LayoutRectangle) => {
    setWidth(evt.width);
  };

  useEffect(() => {
    setPauseTime(0);
    setStartTime(0);
  }, [currentIndex, setStartTime, setPauseTime]);

  useEffect(() => {
    if (index === currentIndex) {
      if (props.pause) {
        const endTime = Date.now();
        setPauseTime(endTime ?? 5);
      }
      if (startTime === 0) {
        setStartTime(Date.now());
      }
    }
  }, [currentIndex, index, props.pause, setPauseTime, setStartTime, startTime]);

  return (
    <View
      onLayout={evt => onLayoutAdded(evt.nativeEvent.layout)}
      style={[
        styles.progressBarContainer,
        {
          backgroundColor: barInActiveColor,
          height: barHeight,
        },
      ]}>
      <Animated.View
        style={[
          styles.progressBarContainer,
          styles.currentBarContainer,
          {
            width: scale,
            backgroundColor: barActiveColor,
            height: barHeight,
          },
        ]}
      />
    </View>
  );
};

export default ProgressBar;
