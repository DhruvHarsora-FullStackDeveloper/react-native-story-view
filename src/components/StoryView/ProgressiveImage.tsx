import React from 'react';
import { Animated, View } from 'react-native';
import styles from './styles';
import type { ProgressiveImageProps } from './types';

const ProgressiveImage = (props: ProgressiveImageProps) => {
  const thumbnailAnimated = new Animated.Value(0.2);
  const { thumbnailSource, imgSource, viewStyle, ...reset } = props;

  const imageAnimated = new Animated.Value(0);

  const handleThumbnailLoad = () => {
    Animated.timing(thumbnailAnimated, {
      toValue: 1,
      useNativeDriver: true,
    }).start();
  };

  const onImageLoad = () => {
    Animated.timing(imageAnimated, {
      toValue: 1,
      useNativeDriver: true,
    }).start();
  };

  return (
    <View style={styles.progressiveImageContainer}>
      <Animated.Image
        {...reset}
        source={thumbnailSource}
        style={[viewStyle, { opacity: thumbnailAnimated }]}
        onLoad={handleThumbnailLoad}
        blurRadius={2}
      />
      <Animated.Image
        {...reset}
        source={imgSource}
        style={[styles.imageOverlay, { opacity: imageAnimated }, viewStyle]}
        onLoad={onImageLoad}
        onLoadEnd={() => props.onImageLoaded && props.onImageLoaded()}
      />
    </View>
  );
};

export default ProgressiveImage;
