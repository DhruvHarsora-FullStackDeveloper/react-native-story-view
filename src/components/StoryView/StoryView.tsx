import * as React from 'react';
import { Text, View } from 'react-native';
import styles from './styles';
import type { StoryViewProps } from './types';

const StoryView = ({ label = 'Story View' }: StoryViewProps) => {
  return (
    <View style={styles.container}>
      <Text>{label}</Text>
    </View>
  );
};

export default StoryView;
