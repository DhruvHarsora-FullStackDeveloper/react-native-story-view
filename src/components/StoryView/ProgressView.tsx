import React, { useCallback, useEffect, useRef } from 'react';
import { Animated } from 'react-native';
import ProgressBar from './ProgressBar';
import styles from './styles';
import { ProgressState } from './types';
import type { ProgressBarsProps } from './types';

const ProgressView = (props: ProgressBarsProps) => {
  const opacity = useRef(new Animated.Value(1)).current;

  useEffect(() => {
    if (props.pause) {
      Animated.timing(opacity, {
        toValue: 0,
        duration: 300,
        useNativeDriver: true,
      }).start();
    } else {
      Animated.timing(opacity, {
        toValue: 1,
        duration: 300,
        useNativeDriver: true,
      }).start();
    }
  }, [opacity, props.pause]);

  const getProgressState = useCallback(
    (i: number) => {
      if (props?.pause) {
        return ProgressState.Paused;
      } else if (i === props.currentIndex) {
        return ProgressState.InProgress;
      } else if (i < props.currentIndex) {
        return ProgressState.Completed;
      }
      return ProgressState.Default;
    },
    [props?.pause, props.currentIndex]
  );

  return (
    <Animated.View style={[styles.progressBarArray]}>
      {props.length.map((i: number, index) => (
        <ProgressBar
          index={index}
          key={i}
          barStyle={props.barStyle}
          duration={props.duration || 3}
          currentIndex={props.currentIndex}
          next={props.next}
          length={props.stories.length}
          active={getProgressState(i)}
          isLoaded={props.isLoaded}
          pause={props.pause}
        />
      ))}
    </Animated.View>
  );
};

export default ProgressView;
