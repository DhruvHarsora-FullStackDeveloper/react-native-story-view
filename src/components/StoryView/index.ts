import StoryView from './StoryView';

export { StoryView };
export type { StoryViewProps } from './types';
