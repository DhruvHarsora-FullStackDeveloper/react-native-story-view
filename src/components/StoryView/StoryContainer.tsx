import React, { useEffect } from 'react';
import {
  KeyboardAvoidingView,
  SafeAreaView,
  TouchableOpacity,
  View
} from 'react-native';
import { Metrics } from '../../theme';
import { useStoryContainer } from './hooks';
import ProgressView from './ProgressView';
import StoryView from './StoryView';
import styles from './styles';
import { ClickPosition, StoryContainerProps } from './types';

const StoryContainer = (props: StoryContainerProps) => {
  const {
    progressIndex,
    isPause,
    isLoaded,
    duration,
    opacity,
    onImageLoaded,
    onVideoLoaded,
    changeStory,
    setLoaded,
    setDuration,
    onArrowClick,
    onStoryPressHold,
    onStoryPressRelease
  } = useStoryContainer(props);

  useEffect(() => {
    setLoaded(false);
    setDuration(
      props.stories?.[progressIndex]?.duration ?? Metrics.defaultDuration
    );
  }, [progressIndex, props.stories, setDuration, setLoaded]);

  const storyViewContent = () => {
    return (
      <>
        <View style={props.containerStyle ?? styles.parentView}>
          <TouchableOpacity
            activeOpacity={1}
            delayLongPress={200}
            onPress={(e: { nativeEvent: any }) => changeStory(e.nativeEvent)}
            onLongPress={onStoryPressHold}
            onPressOut={onStoryPressRelease}>
            <StoryView
              stories={props.stories}
              duration={duration}
              onVideoLoaded={onVideoLoaded}
              onImageLoaded={onImageLoaded}
              progressIndex={progressIndex}
              imageStyle={props.imageStyle}
              pause={isPause}
            />
          </TouchableOpacity>
          {(props?.enableProgress || true) && (
            <View style={[styles.progressView, { opacity }]}>
              <ProgressView
                next={() => onArrowClick(ClickPosition.Right)}
                isLoaded={isLoaded}
                duration={duration}
                pause={!props.enableProgress && isPause}
                stories={props.stories}
                currentIndex={progressIndex}
                barStyle={props.barStyle}
                currentStory={props.stories[progressIndex]}
                length={props.stories.map((_, i) => i)}
                progress={{ id: progressIndex }}
              />
            </View>
          )}
          {props?.headerComponent && (
            <View style={[styles.topView, { opacity }]}>
              {props.headerComponent}
            </View>
          )}
          {props?.customView && (
            <View style={[styles.customView, { opacity }]}>
              {props.customView}
            </View>
          )}
        </View>
        {props?.footerComponent && (
          <View style={[styles.bottomView, { opacity }]}>
            {props.footerComponent}
          </View>
        )}
      </>
    );
  };

  return (
    <SafeAreaView style={styles.container}>
      <KeyboardAvoidingView
        style={styles.container}
        behavior={Metrics.isIOS ? 'padding' : 'height'}>
        {props.visible && storyViewContent()}
      </KeyboardAvoidingView>
    </SafeAreaView>
  );
};

export default StoryContainer;
