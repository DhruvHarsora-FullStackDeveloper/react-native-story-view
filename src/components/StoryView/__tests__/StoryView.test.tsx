import React from 'react';
import { render } from '@testing-library/react-native';
import StoryView from '../StoryView';

jest.useFakeTimers();

describe('CounterButton component', () => {
  it('Match Snapshot', () => {
    const { toJSON } = render(<StoryView />);
    expect(toJSON()).toMatchSnapshot();
  });
});
