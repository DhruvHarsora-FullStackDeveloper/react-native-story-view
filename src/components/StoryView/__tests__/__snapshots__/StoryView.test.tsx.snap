// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CounterButton component Match Snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text>
    Story View
  </Text>
</View>
`;
